File: leetcode/33.search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        
        if (nums.size() == 0)
        {
            return -1;
        }
        
        int start = 0;
        int end = (int)nums.size() - 1;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            
            if (nums[mid] == target)
            {
                return mid;
            }
            
            if (nums[mid] > nums[end])
            {
                if (target >= nums[start] && target < nums[mid])
                {
                    end = mid;
                }
                else
                {
                    start = mid;
                }
                
            }
            else if(nums[mid] < nums[end])
            {
                if (target > nums[mid] && target <= nums[end])
                {
                    start = mid;
                }
                else
                {
                    end = mid;
                }
            }
            else
            {
                end -= 1;
            }
        }
        
        if (nums[start] == target)
        {
            return start;
        }
        
        if (nums[end] == target)
        {
            return end;
        }
        
        return -1;
    }
};

// Interesting solution
// Reference https://leetcode.com/problems/search-in-rotated-sorted-array/discuss/14435/Clever-idea-making-it-simple
class Solution2 {
public:
    int search(vector<int>& nums, int target) {
        
        if (nums.size() == 0)
        {
            return -1;
        }
        
        int start = 0;
        int end = nums.size() - 1;
        while (start + 1 < end) {
            
            int mid = start + (end - start) / 2;
            
            double num = (nums[mid] < nums[0]) == (target < nums[0])
            ? nums[mid]
            : target < nums[0] ? -INFINITY : INFINITY;
            
            if (num < target)
            {
                start = mid;
            }
            else if (num > target)
            {
                end = mid;
            }
            else
            {
                return mid;
            }
        }
        
        if (nums[start] == target)
        {
            return start;
        }
        
        if (nums[end] == target)
        {
            return end;
        }
        
        return -1;
    }
};

File: leetcode/33.search-in-rotated-sorted-array.py
#  Tag: Array, Binary Search
#  Time: O(logN)
#  Space: O(1)
#  Ref: -
#  Note: Rotated;

#  There is an integer array nums sorted in ascending order (with distinct values).
#  Prior to being passed to your function, nums is possibly rotated at an unknown pivot index k (1 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3 and become [4,5,6,7,0,1,2].
#  Given the array nums after the possible rotation and an integer target, return the index of target if it is in nums, or -1 if it is not in nums.
#  You must write an algorithm with O(log n) runtime complexity.
#   
#  Example 1:
#  Input: nums = [4,5,6,7,0,1,2], target = 0
#  Output: 4
#  Example 2:
#  Input: nums = [4,5,6,7,0,1,2], target = 3
#  Output: -1
#  Example 3:
#  Input: nums = [1], target = 0
#  Output: -1
#  
#   
#  Constraints:
#  
#  1 <= nums.length <= 5000
#  -104 <= nums[i] <= 104
#  All values of nums are unique.
#  nums is an ascending array that is possibly rotated.
#  -104 <= target <= 104
#  
#  

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left < right:

            mid = left + (right - left) // 2

            if nums[mid] > nums[right]:
                if nums[right] < target <= nums[mid]:
                    right = mid
                else:
                    left = mid + 1
            else:
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid

        return left if nums[left] == target else -1
    
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        left = 0
        right = n - 1

        while left < right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid

            if target > nums[right]:
                if nums[mid] > nums[right] and nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid
            else:
                if nums[mid] <= nums[right] and nums[mid] >= target:
                    right = mid
                else:
                    left = mid + 1

        return left if nums[left] == target else -1
